File: java/src/priorityQueue/convertArrayIntoHeap.java
package priorityQueue;

import java.util.LinkedList;
import java.util.Scanner;

public class convertArrayIntoHeap {
    private int n;

    private int[] A;

    private LinkedList result = new LinkedList();

    public convertArrayIntoHeap(int n, int[] A) {
        this.n = n;
        this.A = A;
    }

    private int parent(int i) {
        return (int) Math.floor((i - 1) / 2);
    }

    private int leftChild(int i) {
        return i * 2 + 1;
    }

    private int rightChild(int i) {
        return i * 2 + 2;
    }

    private void swap(int i, int j) {
        int tmp = this.A[i];
        this.A[i] = this.A[j];
        this.A[j] = tmp;
        this.result.add(i + " " + j);
    }

    private void shiftUp(int i) {
        while (i > 0) {
            if (this.A[i] < this.A[this.parent(i)])
                this.swap(this.parent(i), i);
            i = this.parent(i);
        }
    }

    public void buildHeap() {
        for (int i = n - 1; i > 0; i--)
            this.shiftUp(i);

        System.out.println(this.result.size());
        this.result.stream().forEach((data) -> System.out.println(data));
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();

        int[] A = new int[n];
        for (int i = 0; i < n; i++) {
            A[i] = scanner.nextInt();
        }

        convertArrayIntoHeap heap = new convertArrayIntoHeap(n, A);
        heap.buildHeap();
    }
}

File: python/priorityQueue/convertArrayIntoHeap.py

from random import *
from heapq import heapify

def parent(i):
    return (i - 1) // 2

def leftChild(i):
    return 2 * i + 1

def rightChild(i):
    return 2 * i + 2

def swap(A, R, i, j):
    A[i], A[j] = A[j], A[i]
    R.append(str(i) + ' ' + str(j))

def shiftUp(A, R, i):
    while i > 0:
        if A[i] < A[parent(i)]:
            swap(A, R, parent(i), i)
        i = parent(i)

def buildHeap(n, A):
    R = []
    for i in range(n - 1, 0, -1):
        shiftUp(A, R, i)
    return R

# input
'''
n = int(input())
A = [*map(int, input().split())]
buildHeap(n, A)
'''

# random value test
N, I= map(int, input().split())

while True:
    n = randint(1, N)
    A = [-1] * n
    for i in range(n):
        A[i] = randint(0, I)

    print('------------------------------------------------------------')
    R = buildHeap(n, A)
    print(len(R))
    for item in R:
        print(item)
    
    if len(R) > 4*n:
        print('error!', len(R), 4*n)
        break

